src/apps/chifra/pkg/rpcClient/contract.go

```diff
@@ -35,3 +37,50 @@ func IsContractAt(chain string, address base.Address, block *types.SimpleNamedBl
 
 	return nil
 }
+
+var deployedCache = make(map[base.Address]base.Blknum)
+
+func GetContractDeployBlock(chain string, address base.Address) (block base.Blknum, err error) {
+	if cached, ok := deployedCache[address]; ok {
+		block = cached
+		return
+	}
+
+	provider := config.GetRpcProvider(chain)
+	latest := BlockNumber(provider)
+
+	if err = IsContractAt(chain, address, &types.SimpleNamedBlock{BlockNumber: latest}); err != nil {
+		return
+	}
+
+	found := sort.Search(int(latest)+1, func(blockNumber int) bool {
+		err := IsContractAt(chain, address, &types.SimpleNamedBlock{BlockNumber: base.Blknum(blockNumber)})
+		return err == nil
+	})
+
+	block = base.Blknum(found)
+	deployedCache[address] = block
```

Comment:
Do we need a mutex?

Reply:
If the code is concurrent, yes. Good catch 👍